modules/dnn/src/cuda/math.cu

```diff
@@ -0,0 +1,210 @@
+
+    namespace detail {
+        template <class T> __device__ T abs(T val);
+        template <> __device__ float abs(float val) { return fabsf(val); }
+        template <> __device__ double abs(double val) { return fabs(val); }
+
+        template <class T> __device__ T exp(T val);
+        template <> __device__ float exp(float val) { return expf(val); }
+        template <> __device__ double exp(double val) { return exp(val); }
+
+        template <class T> __device__ T max(T x, T y);
+        template <> __device__ float max(float x, float y) { return fmaxf(x, y); }
+        template <> __device__ double max(double x, double y) { return fmax(x, y); }
+
+        template <class T> __device__ T min(T x, T y);
+        template <> __device__ float min(float x, float y) { return fminf(x, y); }
+        template <> __device__ double min(double x, double y) { return fmin(x, y); }
+
+        template <class T> __device__ T log(T val);
+        template <> __device__ float log(float val) { return logf(val); }
+        template <> __device__ double log(double val) { return log(val); }
+
+        template <class T> __device__ T tanh(T val);
+        template <> __device__ float tanh(float val) { return tanhf(val); }
+        template <> __device__ double tanh(double val) { return tanh(val); }
+
+        template <class T> __device__ T pow(T val, T exp);
+        template <> __device__ float pow(float val, float exp) { return powf(val, exp); }
+        template <> __device__ double pow(double val, double exp) { return pow(val, exp); }
+
+        template <class T>
+        __device__ T sigmoid(T val) { return T(1) / (1 + exp(-val)); }
+    }
+
+    namespace raw {
+        template <class T>
+        __global__ void abs(view<T> src, span<T> dest) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::abs;
+                dest[i] = abs(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void tanh(view<T> src, span<T> dest) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::tanh;
+                dest[i] = tanh(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void sigmoid(view<T> src, span<T> dest) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::sigmoid;
+                dest[i] = sigmoid(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void bnll(view<T> src, span<T> dest) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::log;
+                using detail::exp;
+                dest[i] = log(1 + exp(-src[i]));
```

Comment:
`log(1+exp(x))` is numerically unstable for large or small values of `x`.  You should use something like this https://github.com/davisking/dlib/blob/master/dlib/dnn/utilities.h#L16 instead.

Consider for example what happens when `x` is very negative.  The addition of 1 and the very small number coming out of exp will underflow to 1 and the result will be exactly 0, which is wrong.  For example, `log(1+exp(-1000))` will yield exactly 0, which isn't right.  It's even worse if you do `log(1+exp(1000))` which will overflow to infinity, but really it should give essentially 1000 as output.

Reply:
How were the bounds decided?

Reply:
I forget how I decided. But I did something sensible :)